Improve type hints on public interface

roberto-arista opened the pull request:
Hey!

I've recently noticed a couple of issues with hints:
- some **kwargs wrongly hinted as `dict[str, something]`
- missing return type hints (`pyright` does infer most of them but `mypy` instead defaults to [`Any`](https://docs.python.org/3/library/typing.html#the-any-type), so better specify them)

I've also left a few comments on the code on hints/validations that I'd like an input from you. I'll try to expose them as discussions here in the PR thread.

I propose to open a separate PR to complete the ruff integration. They can probably go in parallel.

roberto-arista committed:
add missing return type hints

roberto-arista committed:
fix annotations (with note)

roberto-arista committed:
notes

roberto-arista committed:
ruff

roberto-arista committed:
improve hint

roberto-arista committed:
avoid issue with circular import

roberto-arista committed:
avoid issue with circular import

justvanrossum commented on drawBot/drawBotDrawingTools.py:
```diff
@@ -1693,7 +1714,11 @@ def writingDirection(self, direction: str | None):
         self._addInstruction("writingDirection", direction)
 
-    def openTypeFeatures(self, *args: bool | None, **features: bool) -> dict[str, bool]:
+    def openTypeFeatures(
```
Perhaps it's time to remove it. It's for bw compat of an earlier call signature. See the implementation.

roberto-arista replied:
I assume you refer to `openTypeFeatures(None)`. @typemytype what do you think?

justvanrossum commented on drawBot/drawBotDrawingTools.py:
```diff
@@ -1734,7 +1759,9 @@ def listOpenTypeFeatures(self, fontNameOrPath: SomePath | None = None) -> list[s
     listOpenTypeFeatures.__doc__ = FormattedString.listOpenTypeFeatures.__doc__
 
-    def fontVariations(self, *args: None, **axes: float | bool):
+    def fontVariations(
```
`bool`: I think it's for the `resetVariations` case. See the implementation. It's an odd call signature.
`args` here is also for bw compatibility. Perhaps time to remove.

Perhaps this should be rewritten like this:
```python
    def fontVariations(self, *, resetVariations: bool = False, **axes: float) -> dict[str, float]:
        ...
```

I bet the current version predates the "*" feature that says "only keyword args follow". That's the main point: we want to ensure `resetVariations` is passed as a keyword argument.

justvanrossum commented on drawBot/drawBotDrawingTools.py:
```diff
@@ -2203,9 +2250,16 @@ def imagePixelColor(
+        )
 
-    def imageResolution(self, path: SomePath | AppKit.NSImage) -> int:
+    def imageResolution(
```
Maybe do `int | float` instead?

(May I remind you that you yourself authored this annotation? :) )

roberto-arista replied:
it was a while ago 😅

`float` covers `int`, so no need to specify both. Maybe I assumed that an image resolution has to be an integer, from what I understand I was wrong : )

justvanrossum commented on drawBot/drawBotDrawingTools.py:
```diff
@@ -1530,14 +1549,14 @@ def tracking(self, value: float):
         self._addInstruction("baselineShift", value)
 
-    def underline(self, value: str):
+    def underline(self, value: str | None) -> None:  # FIXME could we assert the value in entrance?
```
We'd have the same situation with `lineJoin()` and `lineCap()`. `Literal` sounds good. Is it not too new?

justvanrossum replied:
(Oh, Literal was added in 3.8, we should use it...)

roberto-arista replied:
we could also use it for the `align` keyword of the text functions

justvanrossum commented on drawBot/drawBotDrawingTools.py:
```diff
@@ -338,7 +340,8 @@ def newPage(self, width: str | float | None = None, height: float | None = None)
 
-    def pages(self) -> tuple[DrawBotPage, ...]:
+
+    def pages(self) -> tuple["DrawBotPage", ...]:
```
Perhaps you can use `from __future__ import annotations` instead? That defers evaluation of the annotations, so you don't have to quote the type.

roberto-arista replied:
good idea!

justvanrossum commented on drawBot/drawBotDrawingTools.py:
```diff
@@ -498,7 +496,7 @@ def printImage(self, pdf=None):
             context.printImage(pdf)
 
-    def pdfImage(self):
+    def pdfImage(self):  # FIXME how could we annotate this?
```
It seems it would be `Quartz.PDFDocument | None`, no? Or is `Quartz.PDFDocument` not a usable type? In that case `Any | None` may be better than nothing? Or just `Any`?

roberto-arista replied:
`Quartz.PDFDocument | None` should do!

justvanrossum replied:
`getNSPDFDocument()` can return `None`.

justvanrossum commented on drawBot/drawBotDrawingTools.py:
```diff
@@ -390,7 +388,7 @@ def pages(self):
         return tuple(DrawBotPage(instructionSet) for instructionSet in instructions)
 
-    def saveImage(self, path: SomePath, *args, **options: dict[str, Any]):
+    def saveImage(self, path: SomePath, *args: Any, **options: Any):  # FIXME how could we annotate this one?
```
Looks like it is
```python
    def saveImage(self, ...) -> None | AppKit.NSImage | PIL.ImageFile.ImageFile:
        ...
```
No opinion about how useful that is.

justvanrossum replied:
Or rather `None | list[AppKit.NSImage] | list[PIL.ImageFile.ImageFile]`

roberto-arista replied:
so, the two types cannot be mixed in the same list, right?

justvanrossum replied:
Yes, the code will never return a list with the types mixed.

roberto-arista committed:
defer evalutation of the annotations

roberto-arista committed:
sort imports

roberto-arista committed:
sort imports

roberto-arista committed:
improve hints

roberto-arista committed:
format

roberto-arista committed:
add `None`

roberto-arista committed:
Add `Literal` hints

methods affected:
- `lineCap`
- `underline`
- `strikethrough`
- `writingDirection`
- `text`
- `textBox`
- `textBoxBaselines`

roberto-arista commented:
morning! @typemytype could you have a look at the PR? thanks : )

justvanrossum commented:
I think you should first make sure that the CI doesn't fail. There's currently a mypy error.

roberto-arista commented:
Sure, I'll take care of it. I am curious to know what @typemytype thinks of the open discussions : )

roberto-arista committed:
improve hints

roberto-arista committed:
improve hints

roberto-arista committed:
we can import PIL, so we can get this one back

roberto-arista committed:
Merge branch 'typemytype:master' into master

roberto-arista committed:
improve docs

roberto-arista committed:
improve hints

typemytype committed:
fix test data

a small change in booleanOperation causes this diff

typemytype commented:
fixed the tests

only the `fontVariations` `resetVariations`

roberto-arista committed:
promote resetVariations to explicit keyword argument

this removes the fontVariations(None) deprecated option after 8 years.

roberto-arista committed:
promote `resetFeatures` to explicit keyword argument

openTypeFeatures(None) no longer supported after 8 years of deprecation

roberto-arista committed:
Align all examples to use `newPage` instead of `size`

roberto-arista commented:
I have updated the `fontVariations` to include resetVariations (and mirror the openTypeFeatures example), but then the test is failing...

roberto-arista commented:
I also took the liberty to move all the examples in the docstrings to `newPage()` from `size()`, since I remember it being preferred. And in this way the examples are more consistent. I didn't touch the scripts in the test folder, I can do that as well if you prefer.

justvanrossum commented:
> I also took the liberty to move all the examples in the docstrings to newPage() from size(), since I remember it being preferred

I am not aware of that preference. Either way, such a change should not be part of a PR that intends to add type annotations.

The test failure seems to be that the example script and the expected image are out of sync (script draw three lines, expected image only contains two):

<img width="1045" height="257" alt="image" src="https://github.com/user-attachments/assets/15df3169-2a36-44d7-8f24-af217b71d5ac" />

roberto-arista commented:
> I am not aware of that preference. Either way, such a change should not be part of a PR that intends to add type annotations.

I see your point, but since I've done some corrections to the docstrings in this PR, I thought to add this as well. Type hints add some form of functionalities for IDEs, but in the end they are documentation. I see both falling under the same umbrella. Either way, I can revert it, no problem.

roberto-arista committed:
Revert "Align all examples to use `newPage` instead of `size`"

This reverts commit 924b6c437eb8f6c9f4a7e0478427fada6c78450d.

roberto-arista committed:
align to the other scripts in the folder

roberto-arista committed:
update image

roberto-arista committed:
restore print block

typemytype committed:
fixing test

typemytype committed:
fixing test pdf

typemytype commented:
looks good to me!
